core/src/main/java/io/cucumber/core/runtime/Runtime.java

```diff
@@ -157,12 +163,18 @@ public Builder withEventBus(final EventBus eventBus) {
         }
 
         public Runtime build() {
-            final ObjectFactoryServiceLoader objectFactoryServiceLoader = new ObjectFactoryServiceLoader(classLoader,
-                runtimeOptions);
+            final ObjectFactorySupplier objectFactorySupplier;
+            
+            if(this.objectFactorySupplier != null) {
+                objectFactorySupplier = this.objectFactorySupplier;
+            } else {
+                final ObjectFactoryServiceLoader objectFactoryServiceLoader = new ObjectFactoryServiceLoader(classLoader,
+                    runtimeOptions);
 
-            final ObjectFactorySupplier objectFactorySupplier = runtimeOptions.isMultiThreaded()
-                    ? new ThreadLocalObjectFactorySupplier(objectFactoryServiceLoader)
-                    : new SingletonObjectFactorySupplier(objectFactoryServiceLoader);
+                objectFactorySupplier = runtimeOptions.isMultiThreaded()
+                        ? new ThreadLocalObjectFactorySupplier(objectFactoryServiceLoader)
+                        : new SingletonObjectFactorySupplier(objectFactoryServiceLoader);
```

Comment:
By replacing the  `ObjectFactorySupplier` the `ThreadLocalObjectFactorySupplier` and `SingletonObjectFactorySupplier` are no longer correctly applied. Considering making the `ObjectFactoryServiceLoader` implement the `ObjectFactorySupplier` and so that `ThreadLocalObjectFactorySupplier` and `SingletonObjectFactorySupplier` can be implemented as proper decorators.